cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java

```diff
@@ -78,4 +89,43 @@ private String gatherMetadata(Id.Run run, Id.NamespacedId id) {
     Map<String, String> dataMetadata = businessMetadataStore.getProperties(id);
     return GSON.toJson(dataMetadata);
   }
+
+  private boolean alreadyRegistered(Id.Run run, Id.NamespacedId data, AccessType accessType,
+                                    @Nullable Id.NamespacedId component) {
+    return registered.putIfAbsent(new CheckKey(run, data, accessType, component), true) != null;
+  }
+
+  private static final class CheckKey {
```

Comment:
At this stage, since changing name will require another build. And this is a private class, I'll leave it as-is for now. We can change this when we touch this class again.